### PR TITLE
WIP: doh-stub: Add passthrough DNS functionality

### PR DESCRIPTION
**Pull request by rfinnie**
Add --passthrough-domains and associated options, to allow for a list of domains which should not be sent to the DoH server.

### CODE REVIEW

File: dohproxy/client_protocol.py [hunk 1]
```diff
@@ -25,6 +27,14 @@ def __init__(self, args, logger=None):
...
             self.logger = utils.configure_logger('StubServerProtocol')
         self.client = None
+        self.passthrough_domains = []
+        passthrough_args = getattr(self.args, 'passthrough_domains', '')
+        for domain in passthrough_args.split(','):
+            if not domain:
+                continue
+            if not domain.endswith('.'):
+                domain += '.'
+            self.passthrough_domains.append(dns.name.Name(domain.split('.')))
```
Comment (chantra) [inline dohproxy/client_protocol.py:37]: you can use `argparse`'s `nargs=*` to directly support a list and change that to:

```
self.passthrough_domains = [dns.name.from_text(x) for x in self.args.passthrough_domains]
```

File: dohproxy/client_protocol.py [hunk 2]
```diff
@@ -50,10 +60,20 @@ async def setup_client(self):
     def connection_made(self, transport):
         self.transport = transport
 
+    def is_passthrough_query(self, dnsq):
+        for question in dnsq.question:
```
Comment (chantra) [inline dohproxy/client_protocol.py:64]: I would be tempted to only support single question..... but I think ATM, the proxy pretty much support multiple questions as it does not care about the content really.
I see this being a potential surface of abuse if `passthrough_domains` grows in size but probably fine depending on the use-case and where this is used. Maybe a configurable option to limit how many questions are check would help, with default to 1 being sensible.

File: dohproxy/client_protocol.py [hunk 3]
```diff
@@ -50,10 +60,20 @@ async def setup_client(self):
...
+        return False
+
     def datagram_received(self, data, addr):
         dnsq = dns.message.from_wire(data)
 
-        asyncio.ensure_future(self.make_request(addr, dnsq))
+        if self.is_passthrough_query(dnsq):
+            asyncio.ensure_future(self.resolve(addr, dnsq))
+        else:
+            asyncio.ensure_future(self.make_request(addr, dnsq))
```
Comment (chantra) [inline dohproxy/client_protocol.py:76]: This part will need to be changed based on your other PR #85 .

File: dohproxy/client_protocol.py [hunk 4]
```diff
@@ -145,3 +165,14 @@ async def make_request(self, addr, dnsq):
...
         self.logger.debug('Response trailers: {}'.format(trailers))
+
+    async def resolve(self, addr, dnsq):
+        qid = dnsq.id
+        dnsq.id = 0
+        dnsclient = DNSClient(self.args.passthrough_resolver,
+                              self.args.passthrough_port,
+                              logger=self.logger)
+        dnsr = await dnsclient.query(dnsq, addr)
+        dnsr.id = qid
```
Comment (chantra) [inline dohproxy/client_protocol.py:176]: I don't think you need to care for the `qid` here. `DNSClient._try_query` does that for you (e.g replace with a random one when sending out, setting it back to original value when receiving an answer.

File: dohproxy/stub.py [hunk 5]
```diff
@@ -23,6 +23,24 @@ def parse_args():
         default='::1',
         help='The address the stub should listen on. Default: [%(default)s]'
     )
+    parser.add_argument(
+        '--passthrough-domains',
+        default='',
+        help='Comma-separated list of domains to send to '
+             '--passthrough-resolver. Default: [%(default)s]'
+    )
```
Comment (chantra) [inline dohproxy/stub.py:31]: let's use nargs='*' and `default=[]`

**Comment by rfinnie** [comment]
> I would be interesting in hearing of the use case. I would expect this to be fine _if_ `passthrough_domains` is a small list as otherwise it will become quite costly to iterate over and check if subdomain. That being said, in such case, the data structure used can be improved.

Sure, my use case is my home network with a named private network (snowman.lan, long story, on 10.9.8.0/21).  I set up a remote DoH proxy, locally moved my existing bind9 server to an alternative port, and set up doh-stub.  This is so I can hand off snowman.lan. and 10.in-addr.arpa. to the existing bind9 while sending everything else to DoH.

(It also explains the theme of my other patches: TCP to support EDNS, multiple listen addresses to more closely mimic bind9's interface handling options for a home router setup with multiple addresses per interface, ECS to avoid the theoretical wrath of my friend who works at a major CDN :)

**Comment by rfinnie** [comment]
You know, after reading what I wrote yesterday, I said "wait, what?"  Why am I replacing bind9 with doh-stub and passing private domains back to bind9 on an alternate port?  Why not just keep bind9 on 53 and pass external resolution to doh-stub via "forwarders"?

So yeah. This might still be useful in some situations, but I'm going to mark it WIP, at least until the rest of the merges in my pipeline are addressed (and at least until #85 because of the rebase need).

**Pull request closed without merging**